docs/beman-standard.md

```diff
@@ -0,0 +1,179 @@
+  are prefixed by "RULE:".
+* **Recommendations** should be followed in general, but specific circumstances
+  may make this a less-than-ideal choice. Libraries not following a specific
+  recommendation can still conform to this standard. These entries are prefixed
+  by "RECOMMENDATION:".
+
+## License
+
+**[LICENSE.APPROVED]** REQUIREMENT: All Beman libraries must be licensed
+under an approved license. These are:
+
+1. [Apache License v2.0 with LLVM Exceptions](https://llvm.org/LICENSE.txt)
+2. [Boost Software License 1.0](https://www.boost.org/LICENSE_1_0.txt)
+3. [The MIT License](https://opensource.org/license/mit)
+
+**[LICENSE.BOOST]** RECOMMENDATION: All Beman libraries should be licensed under
+the [Apache License v2.0 with LLVM Exceptions](https://llvm.org/LICENSE.txt).
+
+**[LICENSE.CRITERIA]** REQUIREMENT: All approved licenses must meet the
+following requirements:
+
+1. Simple to read and understand.
+2. Permission without fee to copy, use and modify the software for any
+   use (commercial and non-commercial).
+3. Requires that the license appears on all copies of the software source code.
+4. Must not require that the license appears with executables or other binary
+   uses of the library.
+5. Must not require that the source code be available for execution or other
+   binary uses of the library.
+
+## General
+
+**[LIBRARY_NAMES]** REQUIREMENT: Beman libraries names begin with `Beman.`
+followed by an UpperCamelCase short name.
+
+Examples are Beman.SmartPointer and Beman.SenderReceiver.
+
+**[REPOSITORY_NAME]** RECOMMENDATION: The repository should be named after the
+library name excluding the `Beman.` prefix.
+
+For example, a Beman.SmartPointer library's repository should be named
+`SmartPointer`.
+
+## Top-level
+
+The top-level of a Beman library repository must consist of `CMakeLists.txt`,
+`LICENSE`, and `README.md` files.
+
+**[TOPLEVEL.CMAKE]** REQUIREMENT: There must be a `CMakeLists.txt` at the repository's root
+that builds and tests (via. CTest) the library.
+
+**[TOPLEVEL.LICENSE]** REQUIREMENT: There must be a `LICENSE` at the
+repository's root with the contents of an approved license that covers the
+contents of the repository.
+
+**[TOPLEVEL.README]** REQUIREMENT: There must be a markdown-formatted
+`README.md` at the repository's root that describes the library, explains how
+to build it, and links to further documentation.
+
+## `README.md`
+
+**[README.TITLE]** RECOMMENDATION: The `README.md` should begin with a level 1
+header with the name of the library optionally followed with a ":" and short
+description.
+
+For example:
+
+```markdown
+# Beman.SenderReceiver: Scalable Asychronous Program Building Blocks
+```
+
+**[README.PURPOSE]** RECOMMENDATION: Following the title, the `README.md` should
+contain a one- or two-paragraph summary describing the library's purpose.
+
+**[README.IMPLEMENTS]** RECOMMENDATION: Following the purpose and a newline, the
```

Comment:
Should we also state to not include actual paper code (eg markdown, html, pdf etc)? I think each author can have a separate repo for that, but only include relevant implementation in Beman.

Reply:
If I were maintaining a library paper, I'd rather have everything in one place, personally.

Not that I have capacity for library papers these days, but if I were, I'd be tempted to just do an in-repo markdown file for the paper and convert to pdf when mailing. It might render less elegantlantly than LaTeX, but maybe I don't particularly care. I'd want a normal open source workflow for taking suggestions, and I'd generally be open to additional authors, though.

Reply:
I expect that authors have repos with papers and possibly experiments, at least for ideas that exist at this moment.

Also, I not find OK allowing others changes paper content. The paper is strictly a proposal from a group of authors (which can be extended if they want; otherwise it's a conservative group). I see Beman used only for implementing ideas.

TLDR: My opinion is that keeping papers inside Beman can diverge from authors purpose and from a direction already approved.

Reply:
Instances where an author produces both a paper and its Beman implementation will be common. In this case, a single repository allows for paper/implementation updates to be treated as a single unit of change which is great for reviews.  We can also provide a convenient paper template.

On the flip side, there are cases where one Beman library implements several papers (e.g. Beman.Optional26) or where the implementor is not the paper author.  We need to support those situations as well.

I think what's being proposed in this PR is flexible enough to handle these situations. Perhaps a future PR could provide recommendations where papers sources that cohabitate with a Beman repository should go.

Reply:
> TLDR: My opinion is that keeping papers inside Beman can diverge from authors purpose and from a direction already approved.

I think it's more likely that the impl will diverge if it's kept in a different place.  We see this all the time where the paper changes something -- like the name of interface due to committee bikeshed -- but the implementation isn't updated.  I'd like to see both in the same place to start.

Obviously I'm commenting on a closed PR, but i do expect follow on changes as we learn.

Reply:
> If I were maintaining a library paper, I'd rather have everything in one place, personally.
> 
> Not that I have capacity for library papers these days, but if I were, I'd be tempted to just do an in-repo markdown file for the paper and convert to pdf when mailing. It might render less elegantlantly than LaTeX, but maybe I don't particularly care. I'd want a normal open source workflow for taking suggestions, and I'd generally be open to additional authors, though.

This actually brings up tooling support. M. Park markdown is what many use -- we should look into if he uses 'make targets' or how that's done.  Probably someone here knows.

I use .org mode, and we've been discussing with Steve D more direct support for exporting M. Park markdown.  As it is, I exported to html for my last paper.  One really nice thing about .org is the paper renders in repo nicely: https://github.com/JeffGarland/wg21work/blob/main/monadic_expected/monadic_functions_for_expected.org

Reply:
M. Park Markdown?
[This](https://mpark.github.io/programming/2018/11/16/how-i-format-my-cpp-papers/) seems to indicate M. Park is using whatever pandoc supports

Reply:
yeah it looks like he's using pandoc -- which supports a bunch of markdown flavors and probably allows some sort of configuration of the conversion of his special elements.